README.md

```diff
@@ -1,277 +1,899 @@
+
+**All prices for ADA must be in units of Lovelace.**
+
+##### Creating a One-Way Swap
+
+Swap "writers" must first create a swap by using the `CreateSwap` Redeemer in the following manner:
+
+In order to mint beacons with this redeemer, **all of the following must be true**:
+
+1) The beacons must go to an addres protected by the one-way swap validator script.
+2) The beacons must go to an address using a valid staking credential.
+3) The UTxOs with the beacons must have the proper value:
+    - Exactly two kinds of beacons: pair beacon and offer beacon.
+    - The beacons must correspond to the beacons in the datum.
+    - There must be exactly 1 of each beacon.
+    - No extraneous assets are in the UTxO. ADA is always allowed.
+4) The beacons must be stored with the proper inline `SwapDatum`:
+    - `beaconId` == this policy id.
+    - `pairBeacon` == pair beacon asset name for this swap.
+    - `offerId` == asset_id of the offer asset.
+    - `offerName` == asset_name of the offer asset.
+    - `offerBeacon` == offer beacon asset name for this swap's offer asset.
+    - `askId` == asset_id of the ask asset.
+    - `askName` == asset_name of the ask asset.
+    - `swapPrice` denominator > 0
+    - `swapPrice` > 0
+5) The offer asset and ask assets must be different assets.
+
+Once the beacons are minted to the swap address, the spending script does not allow closing the swap
+*unless* the beacons are being burned. This prevents beacons from being sent to unrelated addresses
+and guarantees the integrity of the off-chain queries.
+
+All swaps must be stored with the proper beacons. Due to this requirement, the absolute minimum
+possible value for a swap UTxO to have is about 2 ADA. This minimum value can be reclaimed upon
+closing the swap so it should be thought of as a deposit. **All open swaps require a deposit of at
+least 2 ADA.** While this behavior is due to the current protocol parameters, it is actually
+a good thing since it helps prevents denial-of-service attacks from having a lot of "zombie swaps"
+open (i.e., UTxOs where there is not enough of the offer asset for other users to swap).
+
+It is possible to open swaps for multiple different trading pairs in a single transaction. The
+minting policy is capable of still ensuring that all beacons are stored properly (with the proper
+value and datum). For example, since each beacon's datum is specific to those beacons, mixing up
+beacons and datums will cause the minting transaction to fail.
+
+The `CreateSwap` allows also burning any beacons to enable composition with the `CloseOrUpdate`
+redeemer. More on this later.
+
+
+##### Closing or Updating a Swap
+
+Open Swap UTxOs can be closed or updated by the address owner (signified by the address' staking
+credential).
+
+The `CloseOrUpdate` redeemer allows the owner to recover the deposit stored with the swap. **In
+order to reclaim the deposit, the beacons must be burned.** As the redeemer name suggests, swaps
```

Comment:
Is it possible to have a eUTxO with only the 2 Ada deposit in it because the swap tokens have been completely spent already?  Does this mean that beacon tokens can remain until users reclaim their 2 Ada deposit?

Reply:
> Is it possible to have a eUTxO with only the 2 Ada deposit in it because the swap tokens have been completely spent already?

This wording isn't accurate since the UTxO should also have the assets that were asked for (eg, 2 ADA + 10 DJED after the USDC was taken). But it is possible for "finished" swaps to still have beacon tokens. These swaps are no longer swappable.

This is not an issue though because, when you query the beacons, you can filter out any UTxOs that do not have the offer asset. Koios allows doing this extra filtering server-side so the burden is not placed on the end user. The `cardano-swaps` CLI uses Koios like this. Blockfrost does not support server-side filtering which is why I did not add support for it in the CLI.